src/core/utils/components/VirtualScroll.tsx

```diff
@@ -100,145 +132,224 @@ export type VirtualScrollProps = {
+        firstChild: firstChildHeight,
+        child:
+          Number(getElementHeightWithMargins(child).toFixed(2)) ??
+          firstChildHeight,
+        lastChild: Number(getElementHeightWithMargins(lastChild).toFixed(2)),
+      };
+    }, [visibleChildren.length]);
+
+    const updateVirtualScroll = React.useCallback(() => {
+      const scrollableContainer =
+        scrollContainer.current ??
+        (parentRef.current?.ownerDocument.scrollingElement as HTMLElement);
+      if (!scrollableContainer) {
+        return;
+      }
+      const start = getNumberOfNodesInHeight(
+        childHeight.current.child,
+        scrollableContainer.scrollTop,
+      );
+      const startIndex = Math.max(0, start - bufferSize);
+      const visibleNodes = getVisibleNodeCount(
+        childHeight.current.child,
         start,
         itemsLength,
         scrollableContainer,
-      ),
-    );
+      );
+      visibleIndex.current = { start: start, end: start + visibleNodes - 1 };
+      setStartNode(startIndex);
+      setVisibleNodeCount(visibleNodes);
 
-    if (!parentRef.current) {
-      return;
-    }
-    parentRef.current.style.transform = `translateY(${getTranslateValue(
-      childHeight.current,
-      startIndex,
-    )}px)`;
-  }, [bufferSize, itemsLength]);
-
-  const removeScrollListener = React.useCallback(() => {
-    if (!onScrollRef.current) {
-      return;
-    }
-    !scrollContainer.current ||
-    scrollContainer.current === parentRef.current?.ownerDocument.body
-      ? parentRef.current?.ownerDocument.removeEventListener(
-          'scroll',
-          onScrollRef.current,
-        )
-      : scrollContainer.current.removeEventListener(
-          'scroll',
-          onScrollRef.current,
-        );
-  }, []);
-
-  // Add event listener to the scrollable container.
-  React.useLayoutEffect(() => {
-    removeScrollListener();
-    onScrollRef.current = updateVirtualScroll;
-    if (
+      if (!parentRef.current) {
+        return;
+      }
+      parentRef.current.style.transform = `translateY(${getTranslateValue(
+        childHeight.current.child,
+        startIndex,
+      )}px)`;
+    }, [bufferSize, itemsLength]);
+
+    const onScroll = React.useCallback(() => {
+      updateVirtualScroll();
+    }, [updateVirtualScroll]);
+
+    const removeScrollListener = React.useCallback(() => {
+      if (!onScrollRef.current) {
+        return;
+      }
       !scrollContainer.current ||
       scrollContainer.current === parentRef.current?.ownerDocument.body
-    ) {
-      parentRef.current?.ownerDocument.addEventListener(
-        'scroll',
-        updateVirtualScroll,
-      );
-    } else {
-      scrollContainer.current.addEventListener('scroll', updateVirtualScroll);
-    }
-    return removeScrollListener;
-  }, [updateVirtualScroll, removeScrollListener]);
-
-  React.useLayoutEffect(() => {
-    updateVirtualScroll();
-  }, [scrollContainerHeight, itemsLength, updateVirtualScroll]);
```

Comment:
Was it intentional to remove `itemsLength`?

Reply:
`itemsLength` is in deps of `updateVirtualScroll`, so should work same.